command/connect/envoy/envoy.go

```diff
@@ -0,0 +1,301 @@
+			"as it has full control over the secrets and config of the proxy.")
+
+	c.flags.BoolVar(&c.bootstrap, "bootstrap", false,
+		"Generate the bootstrap.yaml but don't exec envoy")
+
+	c.flags.StringVar(&c.grpcAddr, "grpc-addr", "",
+		"Set the agent's gRPC address and port (in http(s)://host:port format). "+
+			"Alternatively, you can specify CONSUL_GRPC_ADDR in ENV.")
+
+	c.http = &flags.HTTPFlags{}
+	flags.Merge(c.flags, c.http.ClientFlags())
+	c.help = flags.Usage(help, c.flags)
+}
+
+func (c *cmd) Run(args []string) int {
+	if err := c.flags.Parse(args); err != nil {
+		return 1
+	}
+	passThroughArgs := c.flags.Args()
+
+	// Load the proxy ID and token from env vars if they're set
+	if c.proxyID == "" {
+		c.proxyID = os.Getenv(proxyAgent.EnvProxyID)
+	}
+	if c.sidecarFor == "" {
+		c.sidecarFor = os.Getenv(proxyAgent.EnvSidecarFor)
+	}
+	if c.grpcAddr == "" {
+		c.grpcAddr = os.Getenv(api.GRPCAddrEnvName)
+	}
+	if c.grpcAddr == "" {
+		c.UI.Error("Either -grpc-addr or CONSUL_GRPC_ADDR must be specified")
+		return 1
+	}
+	if c.http.Token() == "" {
+		// Extra check needed since CONSUL_HTTP_TOKEN has not been consulted yet but
+		// calling SetToken with empty will force that to override the
+		if proxyToken := os.Getenv(proxyAgent.EnvProxyToken); proxyToken != "" {
+			c.http.SetToken(proxyToken)
+		}
+	}
+
+	// Setup Consul client
+	client, err := c.http.APIClient()
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Error connecting to Consul agent: %s", err))
+		return 1
+	}
+	c.client = client
+
+	// See if we need to lookup proxyID
+	if c.proxyID == "" && c.sidecarFor != "" {
+		proxyID, err := c.lookupProxyIDForSidecar()
+		if err != nil {
+			c.UI.Error(err.Error())
+			return 1
+		}
+		c.proxyID = proxyID
+	}
+	if c.proxyID == "" {
+		c.UI.Error("No proxy ID specified. One of -proxy-id or -sidecar-for is " +
+			"required")
+		return 1
+	}
+
+	// Generate config
+	bootstrapYaml, err := c.generateConfig()
+	if err != nil {
+		c.UI.Error(err.Error())
+		return 1
+	}
+
+	if c.bootstrap {
+		// Just output it and we are done
+		fmt.Println(bootstrapYaml)
+		return 0
+	}
+
+	// Find Envoy binary
+	binary, err := c.findBinary()
+	if err != nil {
+		c.UI.Error("Couldn't find envoy binary: " + err.Error())
+		return 1
+	}
+
+	// First argument needs to be the executable name.
+
+	// TODO(banks): passing config including an ACL token on command line is jank
+	// - this is world readable. It's easiest thing for now. Temp files are kinda
+	// gross in a different way - we can limit to same-user access which is much
+	// better but we are leaving the ACL secret on disk unencrypted for an
+	// uncontrolled amount of time and in a location the operator doesn't even
+	// know about. Envoy doesn't support reading bootstrap from stdin or ENV
+	envoyArgs := []string{binary, "--config-yaml", bootstrapYaml}
+	envoyArgs = append(envoyArgs, passThroughArgs...)
+
+	// Exec
+	err = syscall.Exec(binary, envoyArgs, os.Environ())
```

Comment:
Does this have cross-compile issues? Specifically: does this work on Windows? I have a tingly feeling we'll need a os-specific function here with build tags and do a subprocess on Windows.

Reply:
It probably does - in an differnt iteration I actually had build tags that explicitly return an error for windows given that Envoy can't build on windows anyway so the best we can do is allow generating a bootstrap there for execution elsewhere.

I didn't go with that approach which attempted to use named pipes to avoid credentials in the ALI argument as it just didn't work without a lot more machinery and effort and lost the platform-specifc error message in the process.

I took it out thinking that this way it's at least possible this could work if someone makes an envoy-compatible proxy for windows or ports Envoy et. but I didn't think too hard about is `syscall.Exec` actually behaves the same way there. I think for now just limiting to platforms that actually have an envoy implementation is reasonable for the run part of this command.

Reply:
@rboyer is looking at a better Exec approach and will handle platform differences so I'll land this as is and we can fix that before launch.